add code to replace AddMatMat with AddMatMatBatched in Convolutional1…

freewym opened the pull request:
…dComponent

Hi Dan,

I implemented AddMatBlocks() in cu-matrix.cc and its corresponding custon kernel function, with no support for transpose.
I added a test in cu-matrix-test.cc for testing AddMatBlocks().

valgrind and cuda-memcheck both indicate no leaks on cu-matrix-test and nnet-component.

I left bias_grad.AddRowSumMat() in the for-loop because I am still not clear how to do it in one call given the inconsistent dims between bias_grad and cols of out_deriv. Could you please clarify it?

BTW, there is a member function AddMatBlock() in CuMatrixBase, which is easily confused with  AddMatBlocks(). Would it be better to propose another name for AddMatBlocks?

The github said "Can't automatically merge" with kaldi-asr/kaldi/master. Is it because there exists some line conflicts?

Thanks,
Yiming

naxingyu commented:
You may get help from here.
https://help.github.com/articles/resolving-a-merge-conflict-from-the-command-line/

It usually happen when your commit and another commit are changing the 
same lines of code.

On 09/11/2015 09:28 AM, Yiming Wang wrote:
>
> …dComponent
>
> Hi Dan,
>
> I implemented AddMatBlocks() in cu-matrix.cc and its corresponding 
> custon kernel function, with no support for transpose.
> I added a test in cu-matrix-test.cc for testing AddMatBlocks().
>
> valgrind and cuda-memcheck both indicate no leaks on cu-matrix-test 
> and nnet-component.
>
> I leave bias_grad.AddRowSumMat() in the for loop because I am still 
> not clear how to do it in one call given the inconsistent dims between 
> bias_grad and cols of out_deriv. Could you please clarify it?
>
> BTW, there is a member function AddMatBlock() in CuMatrixBase, which 
> is easily confused with AddMatBlocks(). Is it better to propose 
> another name for AddMatBlocks?
>
> The github said "Can't aytomatically merge" with 
> kaldi-asr/kaldi/master. Is it because there exists some line conflicts?
>
> Thanks,
> Yiming
>
> ------------------------------------------------------------------------
>
>
>         You can view, comment on, or merge this pull request online at:
>
> https://github.com/kaldi-asr/kaldi/pull/132
>
>
>         Commit Summary
>
>   * add code to replace AddMatMat with AddMatMatBatched in
>     Convolutional1dComponent
>
>
>         File Changes
>
>   * *M* src/cudamatrix/cu-kernels-ansi.h
>     <https://github.com/kaldi-asr/kaldi/pull/132/files#diff-0> (2)
>   * *M* src/cudamatrix/cu-kernels.cu
>     <https://github.com/kaldi-asr/kaldi/pull/132/files#diff-1> (44)
>   * *M* src/cudamatrix/cu-kernels.h
>     <https://github.com/kaldi-asr/kaldi/pull/132/files#diff-2> (2)
>   * *M* src/cudamatrix/cu-matrix-test.cc
>     <https://github.com/kaldi-asr/kaldi/pull/132/files#diff-3> (24)
>   * *M* src/cudamatrix/cu-matrix.cc
>     <https://github.com/kaldi-asr/kaldi/pull/132/files#diff-4> (37)
>   * *M* src/cudamatrix/cu-matrix.h
>     <https://github.com/kaldi-asr/kaldi/pull/132/files#diff-5> (7)
>   * *M* src/nnet2/nnet-component.cc
>     <https://github.com/kaldi-asr/kaldi/pull/132/files#diff-6> (100)
>
>
>         Patch Links:
>
>   * https://github.com/kaldi-asr/kaldi/pull/132.patch
>   * https://github.com/kaldi-asr/kaldi/pull/132.diff
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/kaldi-asr/kaldi/pull/132>.
>

freewym commented:
@naxingyu Thanks.

danpovey commented:
Regarding "I leave bias_grad.AddRowSumMat() in the for loop because I am
still not clear how to do it in one call given the inconsistent dims
between bias_grad and cols of out_deriv. Could you please clarify it?"--

I mean you should create a temporary matrix with the same number of rows as
out_deriv, but the correct number of columns (same as dim of bias_grad),
and use AddMatBlocks() to create the temporary matrix.  You'd have to make
AddMatBlocks work if the num-cols differs.  (You could make it work even if
both num-cols and num-rows differ, by having two loops in the code).

"BTW, there is a member function AddMatBlock() in CuMatrixBase, which is
easily confused with AddMatBlocks(). Is it better to propose another name
for AddMatBlocks?"

I know about that... I couldn't think of a better alternative.

Dan




On Thu, Sep 10, 2015 at 10:00 PM, Yiming Wang <notifications@github.com>
wrote:

> @naxingyu <https://github.com/naxingyu> Thanks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/132#issuecomment-139428994>.
>

freewym commented:
I have updated the call of bias_grad.AddRowSumMat() , as long as made AddMatBlock() to support both row blocks and col blocks. These 2 commits have be squashed into a single one.

Yiming

danpovey commented on src/nnet2/nnet-component.cc:
```diff
@@ -4170,13 +4211,42 @@ void Convolutional1dComponent::Update(const CuMatrixBase<BaseFloat> &in_value,
+  
+  //allocate a single large matrix holding the smaller matrices from 
+  //the vector container filters_grad_batch
+  CuMatrix<BaseFloat> *filters_grad_blocks_batch = new CuMatrix<BaseFloat>(
```
there no need to allocate this on the heap.

danpovey commented:
Please make sure your AddMatBatch has a speed-test routine.
Also, let me know when you think it's ready to merge.
Dan


On Fri, Sep 11, 2015 at 1:02 AM, Yiming Wang <notifications@github.com>
wrote:

> I have updated the call of bias_grad.AddRowSumMat() , as long as made
> AddMatBlock() to support both row blocks and col blocks. These 2 commits
> have be squashed into a single one.
>
> Yiming
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/132#issuecomment-139453072>.
>

danpovey commented:
Sorry I mean  AddMatBlocks
Dan


On Fri, Sep 11, 2015 at 1:09 AM, Daniel Povey <dpovey@gmail.com> wrote:

> Please make sure your AddMatBatch has a speed-test routine.
> Also, let me know when you think it's ready to merge.
> Dan
>
>
> On Fri, Sep 11, 2015 at 1:02 AM, Yiming Wang <notifications@github.com>
> wrote:
>
>> I have updated the call of bias_grad.AddRowSumMat() , as long as made
>> AddMatBlock() to support both row blocks and col blocks. These 2 commits
>> have be squashed into a single one.
>>
>> Yiming
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kaldi-asr/kaldi/pull/132#issuecomment-139453072>.
>>
>
>

freewym commented:
Hi Dan,

I have pushed a single commit. Let me know if there are any problems.

According to the speed test, AddMatBlocks improves over AddMat on GPU (I removed the speed test for AddMat in the final commit):

LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 16 numRowBlocks = 4 numColBlocks = 4, speed was 0.000310383 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 32 numRowBlocks = 4 numColBlocks = 4, speed was 0.00175293 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 64 numRowBlocks = 4 numColBlocks = 4, speed was 0.0118271 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 128 numRowBlocks = 4 numColBlocks = 4, speed was 0.0301476 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 256 numRowBlocks = 4 numColBlocks = 4, speed was 0.111097 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 512 numRowBlocks = 4 numColBlocks = 4, speed was 0.438642 gigaflops.
LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For CuMatrix::AddMat<double>, for dim = 1024 numRowBlocks = 4 numColBlocks = 4, speed was 1.71881 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 16 numRowBlocks = 4numColBlocks = 4, speed was 0.00687876 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 32 numRowBlocks = 4 numColBlocks = 4, speed was 0.0460843 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 64 numRowBlocks = 4 numColBlocks = 4, speed was 0.186284 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 128 numRowBlocks = 4 numColBlocks = 4, speed was 0.752389 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 256 numRowBlocks = 4 numColBlocks = 4, speed was 2.46113 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 512 numRowBlocks = 4 numColBlocks = 4, speed was 4.26382 gigaflops.
LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For CuMatrix::AddMatBlocks<double>, for dim = 1024 numRowBlocks = 4 numColBlocks = 4, speed was 7.09727 gigaflops.

Yiming

danpovey commented:
OK, very good.  But please make sure the TestCuMatrixAddMat() test code
gets made the same as it was before, when you check in.  I want to keep
that as a test for that one routine.
Dan


On Fri, Sep 11, 2015 at 11:29 AM, Yiming Wang <notifications@github.com>
wrote:

> Hi Dan,
>
> I have pushed a single commit. Let me know if there are any problems.
>
> According to the speed test, AddMatBlocks improves over AddMat on GPU (I
> removed the speed test for AddMat in the final commit):
>
> LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For
> CuMatrix::AddMat, for dim = 16 numRowBlocks = 4 numColBlocks = 4, speed was
> 0.000310383 gigaflops.
> LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For
> CuMatrix::AddMat, for dim = 32 numRowBlocks = 4 numColBlocks = 4, speed was
> 0.00175293 gigaflops.
> LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For
> CuMatrix::AddMat, for dim = 64 numRowBlocks = 4 numColBlocks = 4, speed was
> 0.0118271 gigaflops.
> LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For
> CuMatrix::AddMat, for dim = 128 numRowBlocks = 4 numColBlocks = 4, speed
> was 0.0301476 gigaflops.
> LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For
> CuMatrix::AddMat, for dim = 256 numRowBlocks = 4 numColBlocks = 4, speed
> was 0.111097 gigaflops.
> LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For
> CuMatrix::AddMat, for dim = 512 numRowBlocks = 4 numColBlocks = 4, speed
> was 0.438642 gigaflops.
> LOG (TestCuMatrixAddMat():cu-matrix-speed-test.cc:63) For
> CuMatrix::AddMat, for dim = 1024 numRowBlocks = 4 numColBlocks = 4, speed
> was 1.71881 gigaflops.
> LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For
> CuMatrix::AddMatBlocks, for dim = 16 numRowBlocks = 4numColBlocks = 4,
> speed was 0.00687876 gigaflops.
> LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For
> CuMatrix::AddMatBlocks, for dim = 32 numRowBlocks = 4 numColBlocks = 4,
> speed was 0.0460843 gigaflops.
> LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For
> CuMatrix::AddMatBlocks, for dim = 64 numRowBlocks = 4 numColBlocks = 4,
> speed was 0.186284 gigaflops.
> LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For
> CuMatrix::AddMatBlocks, for dim = 128 numRowBlocks = 4 numColBlocks = 4,
> speed was 0.752389 gigaflops.
> LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For
> CuMatrix::AddMatBlocks, for dim = 256 numRowBlocks = 4 numColBlocks = 4,
> speed was 2.46113 gigaflops.
> LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For
> CuMatrix::AddMatBlocks, for dim = 512 numRowBlocks = 4 numColBlocks = 4,
> speed was 4.26382 gigaflops.
> LOG (TestCuMatrixAddMatBlocks():cu-matrix-speed-test.cc:82) For
> CuMatrix::AddMatBlocks, for dim = 1024 numRowBlocks = 4 numColBlocks = 4,
> speed was 7.09727 gigaflops.
>
> Yiming
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/132#issuecomment-139576273>.
>

freewym commented:
Done.

danpovey commented:
There are conflicts in the pull request-- can you please attempt to rebase
against master, e.g.

 git rebase upstream/master

and fix any conflicts?  Get someone who knows git to help you.

Dan


On Fri, Sep 11, 2015 at 1:43 PM, Yiming Wang <notifications@github.com>
wrote:

> Done.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/132#issuecomment-139610375>.
>

freewym commented:
I figured out: the conflicts lines are in two files.
1. cu-matrix.cc. I removed line 896 and added back at line 931
2. cu-matrix-speed-test.cc: it seems that you added TestCuMatrixResize at line 748/790 after my last checkout. I retained this one.

Ran the test after solving the conflicts. Passed. Checked in the new commit.

danpovey commented:
You implemented the transposed version of AddMatBLocks, as far as I can
see, but I don't see the test code testing that.
Dan


On Fri, Sep 11, 2015 at 5:26 PM, Yiming Wang <notifications@github.com>
wrote:

> I figured out: the conflicts lines are in two files.
> 1. cu-matrix.cc. I removed line 896 and added back at line 931
> 2. cu-matrix-speed-test.cc: it seems that you added TestCuMatrixResize at
> line 748/790 after my last checkout. I retained this one.
>
> Ran the test after solving the conflicts. Passed. Checked in the new
> commit.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/132#issuecomment-139664802>.
>

freewym commented:
I didn't implement it. The function  _add_mat_blocks_trans is actually identical to the no-trans version _add_mat_blocks in cu-kernels.cu. I did this because I noticed that _add_mat and _add_mat_trans seem identical.

danpovey commented:
But your code does support it-- it shouldn't silently give wrong results if
it doesn't support something, it should crash with an error message.
But actually, look more carefully at where index_src is declared-- those
functions are different.
It would be better to implement and test it.
Dan


On Fri, Sep 11, 2015 at 5:46 PM, Yiming Wang <notifications@github.com>
wrote:

> I didn't implement it. The function _add_mat_blocks_trans is actually
> identical to the no-trans version _add_mat_blocks in cu-kernels.cu. I did
> this because I noticed that _add_mat and _add_mat_trans seem identical.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/132#issuecomment-139668642>.
>

freewym commented:
Sorry. I will fix/implement that.

freewym commented:
Please check the new one. The trans version of AddMatBlocks passed the test.

Yiming

danpovey commented:
Thanks!
Merging.
Dan

danpovey commented:
@naxingyu, if you have a test setup for the convolutional stuff now might be a good time to re-test.

naxingyu commented:
OK. Thanks.

On 09/12/2015 08:13 AM, Daniel Povey wrote:
>
> @naxingyu <https://github.com/naxingyu>, if you have a test setup for 
> the convolutional stuff now might be a good time to re-test.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/kaldi-asr/kaldi/pull/132#issuecomment-139691021>.
>
